File: backend/app/engine/date_math.py
"""Date arithmetic and year-fraction calculation supporting Excel YEARFRAC conventions.

Conventions supported:
- Basis 0: US 30/360 (NASD)
- Basis 1: Actual/Actual (Excel convention: actual days / average days in spanned years)
- Basis 2: Actual/360
- Basis 3: Actual/365
- Basis 4: European 30/360
"""

from datetime import date, datetime
from typing import Union

DateLike = Union[str, date, datetime]

BASIS_NAMES = {
    0: "US 30/360",
    1: "Actual/Actual",
    2: "Actual/360",
    3: "Actual/365",
    4: "European 30/360"
}


def parse_date(d: DateLike) -> date:
    """Parse date-like object to datetime.date in UTC/naive format."""
    if isinstance(d, datetime):
        return d.date()
    if isinstance(d, date):
        return d
    if isinstance(d, str):
        # Support YYYY-MM-DD or ISO strings
        s = d.strip()
        if "T" in s:
            s = s.split("T")[0]
        return datetime.strptime(s, "%Y-%m-%d").date()
    raise ValueError(f"Cannot parse date: {d}")


def is_leap_year(y: int) -> bool:
    """Check if year is a leap year according to Gregorian rules."""
    return (y % 4 == 0) and (y % 100 != 0 or y % 400 == 0)


def days_between(d1: date, d2: date) -> int:
    """Actual calendar days between two dates."""
    return (d2 - d1).days


def year_fraction(d1_raw: DateLike, d2_raw: DateLike, basis: int = 1) -> float:
    """Compute year fraction between two dates adhering to the specified day-count basis.

    Preserves exact Excel and reference JS implementation semantics.
    """
    if not d1_raw or not d2_raw:
        return 0.0

    d1 = parse_date(d1_raw)
    d2 = parse_date(d2_raw)

    sign = 1.0
    if d2 < d1:
        d1, d2 = d2, d1
        sign = -1.0

    days = days_between(d1, d2)
    if days == 0:
        return 0.0

    basis = int(basis)
    if basis == 2:
        # Actual/360
        return sign * (days / 360.0)

    if basis == 3:
        # Actual/365
        return sign * (days / 365.0)

    if basis in (0, 4):
        # 30/360 conventions
        y1, m1, dd1 = d1.year, d1.month, d1.day
        y2, m2, dd2 = d2.year, d2.month, d2.day

        if basis == 0:
            # US 30/360
            if dd1 == 31:
                dd1 = 30
            if dd2 == 31 and dd1 >= 30:
                dd2 = 30
        else:
            # European 30/360
            dd1 = min(dd1, 30)
            dd2 = min(dd2, 30)

        num_days = (y2 - y1) * 360 + (m2 - m1) * 30 + (dd2 - dd1)
        return sign * (num_days / 360.0)

    # Basis 1: Actual/Actual (Excel YEARFRAC: actual days / average days in calendar years spanned)
    y1, y2 = d1.year, d2.year
    years_count = y2 - y1 + 1
    total_year_days = sum(366 if is_leap_year(y) else 365 for y in range(y1, y2 + 1))
    avg_year_days = total_year_days / float(years_count)
    return sign * (days / avg_year_days)

File: backend/src/main/java/com/example/contingentanalysis/domain/date/DateMath.java
package com.example.contingentanalysis.domain.date;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

public final class DateMath {

    public static final Map<Integer, String> BASIS_NAMES;

    static {
        Map<Integer, String> map = new LinkedHashMap<>();
        map.put(0, "US 30/360");
        map.put(1, "Actual/Actual");
        map.put(2, "Actual/360");
        map.put(3, "Actual/365");
        map.put(4, "European 30/360");
        BASIS_NAMES = Collections.unmodifiableMap(map);
    }

    private DateMath() {
    }

    public static LocalDate parseDate(String d) {
        if (d == null || d.trim().isEmpty()) {
            throw new IllegalArgumentException("Cannot parse empty date");
        }
        String s = d.trim();
        if (s.contains("T")) {
            s = s.split("T")[0];
        }
        return LocalDate.parse(s, DateTimeFormatter.ISO_LOCAL_DATE);
    }

    public static boolean isLeapYear(int y) {
        return (y % 4 == 0) && (y % 100 != 0 || y % 400 == 0);
    }

    public static long daysBetween(LocalDate d1, LocalDate d2) {
        return ChronoUnit.DAYS.between(d1, d2);
    }

    public static double yearFraction(String d1Raw, String d2Raw, int basis) {
        if (d1Raw == null || d1Raw.trim().isEmpty() || d2Raw == null || d2Raw.trim().isEmpty()) {
            return 0.0;
        }

        LocalDate d1 = parseDate(d1Raw);
        LocalDate d2 = parseDate(d2Raw);

        double sign = 1.0;
        if (d2.isBefore(d1)) {
            LocalDate tmp = d1;
            d1 = d2;
            d2 = tmp;
            sign = -1.0;
        }

        long days = daysBetween(d1, d2);
        if (days == 0) {
            return 0.0;
        }

        if (basis == 2) {
            // Actual/360
            return sign * (days / 360.0);
        }

        if (basis == 3) {
            // Actual/365
            return sign * (days / 365.0);
        }

        if (basis == 0 || basis == 4) {
            // 30/360 conventions
            int y1 = d1.getYear();
            int m1 = d1.getMonthValue();
            int dd1 = d1.getDayOfMonth();

            int y2 = d2.getYear();
            int m2 = d2.getMonthValue();
            int dd2 = d2.getDayOfMonth();

            if (basis == 0) {
                // US 30/360
                if (dd1 == 31) {
                    dd1 = 30;
                }
                if (dd2 == 31 && dd1 >= 30) {
                    dd2 = 30;
                }
            } else {
                // European 30/360
                dd1 = Math.min(dd1, 30);
                dd2 = Math.min(dd2, 30);
            }

            int numDays = (y2 - y1) * 360 + (m2 - m1) * 30 + (dd2 - dd1);
            return sign * (numDays / 360.0);
        }

        // Basis 1: Actual/Actual (Excel YEARFRAC: actual days / average days in calendar years spanned)
        int y1 = d1.getYear();
        int y2 = d2.getYear();
        int yearsCount = y2 - y1 + 1;
        long totalYearDays = 0;
        for (int y = y1; y <= y2; y++) {
            totalYearDays += isLeapYear(y) ? 366 : 365;
        }
        double avgYearDays = (double) totalYearDays / (double) yearsCount;
        return sign * (days / avgYearDays);
    }
}

